File: modules/desktop/frontend/src/lib/memory.js
export const emptyMemoryDraft = {
  scope: 'project',
  kind: 'fact',
  status: 'active',
  title: '',
  content: '',
  confidence: 'medium',
};

export function normalizeMemoryRecord(item) {
  return {
    id: item.id,
    scope: item.scope || 'project',
    kind: item.kind || 'fact',
    status: item.status || 'active',
    title: item.title || item.id,
    content: item.content || '',
    confidence: item.confidence || 'medium',
    workspaceRoot: item.workspace_root || '',
    sessionId: item.session_id || '',
    runId: item.run_id || '',
    source: item.source || 'user',
    metadata: item.metadata || {},
    createdAt: item.created_at,
    updatedAt: item.updated_at,
    deletedAt: item.deleted_at,
  };
}

export function memoryDraftFrom(item) {
  if (!item) return emptyMemoryDraft;
  return {
    scope: item.scope || 'project',
    kind: item.kind || 'fact',
    status: item.status || 'active',
    title: item.title || '',
    content: item.content || '',
    confidence: item.confidence || 'medium',
  };
}

export function memoryCreatePayload(input, context = {}) {
  const scope = input.scope || 'project';
  return stripEmpty({
    scope,
    kind: input.kind || 'fact',
    status: input.status || 'active',
    title: input.title,
    content: input.content,
    confidence: input.confidence || 'medium',
    workspace_root: scope === 'project' ? context.workspaceRoot : '',
    session_id: scope === 'session' ? context.sessionId : '',
    source: 'user',
  });
}

export function memoryUpdatePayload(input) {
  return stripUndefined({
    scope: input.scope,
    kind: input.kind,
    status: input.status,
    title: input.title,
    content: input.content,
    confidence: input.confidence,
  });
}

export function memoryPreviewPayload(context = {}) {
  return {
    session_id: context.sessionId || '',
    workspace_root: context.workspaceRoot || '',
    input: context.input || '',
  };
}

export function memoryListQuery({ scope, status, workspaceRoot, sessionId } = {}) {
  const params = new URLSearchParams();
  if (scope) params.set('scope', scope);
  if (status && status !== 'active') params.set('status', status);
  if (scope === 'project' && workspaceRoot) params.set('workspace_root', workspaceRoot);
  if (scope === 'session' && sessionId) params.set('session_id', sessionId);
  const query = params.toString();
  return query ? `/api/v1/memory?${query}` : '/api/v1/memory';
}

function stripUndefined(value) {
  return Object.fromEntries(Object.entries(value).filter(([, item]) => item !== undefined));
}

function stripEmpty(value) {
  return Object.fromEntries(Object.entries(value).filter(([, item]) => item !== undefined && item !== ''));
}

File: modules/gateway/internal/gateway/repository/memory.go
package repository

import (
	"fmt"
	"sync/atomic"
	"time"

	"gorm.io/gorm"

	"redpanda/gateway/internal/gateway/model"
)

var memoryIDCounter atomic.Uint64

type MemoryRepository struct {
	db *gorm.DB
}

type MemoryListFilter struct {
	Scope         string
	WorkspaceRoot string
	SessionID     string
	Status        string
	Limit         int
}

type MemoryUpdate struct {
	ID           string
	Scope        *string
	Kind         *string
	Status       *string
	Title        *string
	Content      *string
	Confidence   *string
	MetadataJSON *string
}

func NewMemoryRepository(db *gorm.DB) MemoryRepository {
	return MemoryRepository{db: db}
}

func (r MemoryRepository) Create(record model.MemoryRecord) (model.MemoryRecord, error) {
	now := time.Now().UTC()
	if record.ID == "" {
		record.ID = fmt.Sprintf("mem_%d_%d", now.UnixNano(), memoryIDCounter.Add(1))
	}
	if record.Status == "" {
		record.Status = "active"
	}
	if record.Status == "deleted" && record.DeletedAt == nil {
		record.DeletedAt = &now
	}
	record.CreatedAt = now
	record.UpdatedAt = now
	return record, r.db.Create(&record).Error
}

func (r MemoryRepository) Get(id string) (model.MemoryRecord, error) {
	var row model.MemoryRecord
	err := r.db.First(&row, "id = ?", id).Error
	return row, err
}

func (r MemoryRepository) List(filter MemoryListFilter) ([]model.MemoryRecord, error) {
	limit := normalizeMemoryLimit(filter.Limit)
	query := r.db.Model(&model.MemoryRecord{})
	if filter.Status == "" {
		query = query.Where("status = ? AND deleted_at IS NULL", "active")
	} else if filter.Status != "all" {
		query = query.Where("status = ?", filter.Status)
		if filter.Status != "deleted" {
			query = query.Where("deleted_at IS NULL")
		}
	}
	if filter.Scope != "" {
		query = query.Where("scope = ?", filter.Scope)
	}
	if filter.WorkspaceRoot != "" {
		query = query.Where("workspace_root = ?", filter.WorkspaceRoot)
	}
	if filter.SessionID != "" {
		query = query.Where("session_id = ?", filter.SessionID)
	}
	var rows []model.MemoryRecord
	err := query.Order("updated_at desc").Limit(limit).Find(&rows).Error
	return rows, err
}

func (r MemoryRepository) Update(input MemoryUpdate) (model.MemoryRecord, error) {
	current, err := r.Get(input.ID)
	if err != nil {
		return model.MemoryRecord{}, err
	}
	if input.Scope != nil {
		current.Scope = *input.Scope
	}
	if input.Kind != nil {
		current.Kind = *input.Kind
	}
	if input.Status != nil {
		current.Status = *input.Status
		if current.Status == "deleted" && current.DeletedAt == nil {
			now := time.Now().UTC()
			current.DeletedAt = &now
		}
		if current.Status != "deleted" {
			current.DeletedAt = nil
		}
	}
	if input.Title != nil {
		current.Title = *input.Title
	}
	if input.Content != nil {
		current.Content = *input.Content
	}
	if input.Confidence != nil {
		current.Confidence = *input.Confidence
	}
	if input.MetadataJSON != nil {
		current.MetadataJSON = *input.MetadataJSON
	}
	current.UpdatedAt = time.Now().UTC()
	return current, r.db.Save(&current).Error
}

func (r MemoryRepository) Delete(id string) (model.MemoryRecord, error) {
	now := time.Now().UTC()
	if err := r.db.Model(&model.MemoryRecord{}).
		Where("id = ? AND deleted_at IS NULL", id).
		Updates(map[string]any{
			"status":     "deleted",
			"deleted_at": &now,
			"updated_at": now,
		}).Error; err != nil {
		return model.MemoryRecord{}, err
	}
	return r.Get(id)
}

func (r MemoryRepository) SelectForRun(sessionID string, workspaceRoot string, limit int) ([]model.MemoryRecord, error) {
	limit = normalizeMemoryLimit(limit)
	query := r.db.Model(&model.MemoryRecord{}).
		Where("status = ? AND deleted_at IS NULL", "active")
	if sessionID != "" && workspaceRoot != "" {
		query = query.Where("(scope = ? AND session_id = ?) OR (scope = ? AND workspace_root = ?)", "session", sessionID, "project", workspaceRoot)
	} else if sessionID != "" {
		query = query.Where("scope = ? AND session_id = ?", "session", sessionID)
	} else if workspaceRoot != "" {
		query = query.Where("scope = ? AND workspace_root = ?", "project", workspaceRoot)
	} else {
		return nil, nil
	}
	var rows []model.MemoryRecord
	err := query.
		Order(scopePrioritySQL()).
		Order("updated_at desc").
		Limit(limit).
		Find(&rows).Error
	return rows, err
}

func normalizeMemoryLimit(limit int) int {
	if limit <= 0 || limit > 200 {
		return 100
	}
	return limit
}

func scopePrioritySQL() string {
	return "CASE scope WHEN 'session' THEN 0 WHEN 'project' THEN 1 WHEN 'user' THEN 2 ELSE 3 END"
}
